### ISSUE TITLE
civetweb for non-mmu (Linux) system

### ISSUE THREAD

**Issue by jaapcrezee** [civetweb/civetweb#980]
I am asked to help with the development on a Blackfin (MMU-less CPU)  Linux system. I ran into civetweb and gave it a try. I needed to patch `fork` to `vfork` to get it compiling using our toolchain. It seems to be running okay on our target including running cgi's (I don't care about the performance).
Would this project be open to accepting a patch to allow compile time configuration of fork vs vfork using a CMake option? Or maybe there are alternative solutions?
If yes, I could fork this project and prepare a merge request.
If no, I will most likely fork anyway but hard-code my fork to vfork patch without any configuration there.

**Comment by bel2125** [comment]
Yes a patch will be accepted. We always had some pre-processor defines for embedded systems where some API functions are not available. Did you check if there is already an acceptable define, e.g., NO_CGI if you do not have different processes anyway - or don't plan to use CGI?

**Comment by jaapcrezee** [comment]
NO_CGI is not acceptable for me as I am actually using CGI's in this system.
I'll try to make an initial patch for you to review.

**Comment by bel2125** [comment]
ok.

I have no experience with embedded systems that do allow spawning "processes" as required by CGI, but do not have a memory separation between processes. If it is possible that two identical CGI processes can run at the same time (starting the same executable twice), then everything should be fine.
All the embedded systems I ever used that do not use memory separation also did not have the concept of "processes", so CGI made no sense there.
You wrote above you would use vfork(): CivetWeb is a multithreaded process, this might add some complexity. I never used vfork() in a multithreaded environment, but I used posix_spawn() ... probably with different flags as you would need for your non-mmu System.

Anyway, if you create a patch, I will have a close look at it.
Note that we plan a release in the beginning of April - so depending on some risk estimation, we might merge a patch only after this release.

**Comment by jaapcrezee** [comment]
Good to know, will keep that in mind. I will get back to this, but it can take some time as I am only working for this project/customer a day per week.

**Comment by bel2125** [comment]
I don't know what implementation language you want to use for your CGI scripts.
If it is C/C++ (or callable from C/C++), you might just use `mg_set_request_handler` and call the C code directly from there, since it's anyway in the same address space. ... just an idea.

CGI is also based on the concept of stdin/stdout redirection. I don't know how this concept translates to your case. Your "process" does not have it's own separated memory, but does it have it's own separated file descriptors, in particular separated stdin/stdout? Maybe this isn't a problem at all ... as I said, I don't have any experience with a system like that.

**Comment by bel2125** [comment]
No more questions here since a while --> closing this issue
Feel free to reopen it or create a new issue if there is anything else you need.
